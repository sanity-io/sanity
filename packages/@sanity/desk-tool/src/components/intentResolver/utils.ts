// @todo: remove the following line when part imports has been removed from this file
///<reference types="@sanity/types/parts" />
import {getPublishedId} from 'part:@sanity/base/util/draft-utils'

export function removeDraftPrefix(documentId: string): string {
  const publishedId = getPublishedId(documentId)

  if (publishedId !== documentId) {
    console.warn(
      'Removed unexpected draft id in document link: All links to documents should have the ' +
        '`drafts.`-prefix removed and something appears to have made an intent link to `%s`',
      documentId
    )
  }

  return publishedId
}
