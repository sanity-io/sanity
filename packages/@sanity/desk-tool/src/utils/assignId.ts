import {nanoid} from 'nanoid'

// `WeakMap`s require the first type param to extend `object`
// eslint-disable-next-line @typescript-eslint/ban-types
const randomIdCache = new WeakMap<object, string>()

/**
 * given an object, this function randomly generates an ID and returns it. this
 * result is then saved in a WeakMap so subsequent requests for the same object
 * will receive the same ID
 */
// eslint-disable-next-line @typescript-eslint/ban-types
export function assignId(obj: object): string {
  const cachedValue = randomIdCache.get(obj)
  if (cachedValue) return cachedValue

  const id = nanoid()
  randomIdCache.set(obj, id)
  return id
}
