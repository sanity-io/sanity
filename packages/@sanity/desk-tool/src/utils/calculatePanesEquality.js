import {isEqual} from 'lodash'

// eslint-disable-next-line import/prefer-default-export
export const calculatePanesEquality = (prev = [], next = []) => {
  if (prev === next) {
    return {ids: true, params: true}
  }

  if (prev.length !== next.length) {
    return {ids: false, params: false}
  }

  let paramsDiffer = false
  const idsEqual = prev.every((prevGroup, index) => {
    const nextGroup = next[index]
    if (prevGroup.length !== nextGroup.length) {
      return false
    }

    return prevGroup.every((prevPane, paneIndex) => {
      const nextPane = nextGroup[paneIndex]

      paramsDiffer =
        paramsDiffer ||
        !isEqual(nextPane.params, prevPane.params) ||
        !isEqual(nextPane.payload, prevPane.payload)

      return nextPane.id === prevPane.id
    })
  })

  return {ids: idsEqual, params: !paramsDiffer}
}
