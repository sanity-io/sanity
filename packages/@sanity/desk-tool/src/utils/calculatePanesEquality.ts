import {isEqual} from 'lodash'
import {RouterPanes} from '../types'

export const calculatePanesEquality = (
  prev: RouterPanes = [],
  next: RouterPanes = []
): {ids: boolean; params: boolean} => {
  if (prev === next) {
    return {ids: true, params: true}
  }

  if (prev.length !== next.length) {
    return {ids: false, params: false}
  }

  let paramsDiffer = false
  const idsEqual = prev.every((prevGroup, index) => {
    const nextGroup = next[index]
    if (prevGroup.length !== nextGroup.length) {
      return false
    }

    return prevGroup.every((prevPane, paneIndex) => {
      const nextPane = nextGroup[paneIndex]

      paramsDiffer =
        paramsDiffer ||
        !isEqual(nextPane.params, prevPane.params) ||
        !isEqual(nextPane.payload, prevPane.payload)

      return nextPane.id === prevPane.id
    })
  })

  return {ids: idsEqual, params: !paramsDiffer}
}
