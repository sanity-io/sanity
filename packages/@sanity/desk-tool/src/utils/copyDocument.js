import {isObject, omit} from 'lodash'

function copyReference(reference) {
  return Object.assign({}, reference)
}

function copyObject(object, options = {}) {
  return Object.keys(object).reduce((copy, key) => {
    copy[key] = copyAny(object[key], options)
    return copy
  }, {})
}

function copyArray(array, options = {}) {
  return array.map(item => copyAny(item, options)).filter(Boolean)
}

const COPY_AS_IS = ['number', 'string', 'boolean']

function isReference(value) {
  return (
    value._type === 'reference' ||
    // should not happen as all references should have _type === 'reference'
    (!('_type' in value) && '_ref' in value)
  )
}

function copyAny(value, options) {
  if (Array.isArray(value)) {
    return copyArray(value, options)
  }
  const type = typeof value
  if (COPY_AS_IS.includes(type)) {
    return value
  }
  if (!value) {
    return value
  }

  if (isObject(value)) {
    if (isReference(value)) {
      return options.excludeReferences ? undefined : copyReference(value)
    }
    return copyObject(value, options)
  }
  return value
}

export default function copyDocument(doc, options = {}) {
  return copyAny(omit(doc, '_id'), options)
}
