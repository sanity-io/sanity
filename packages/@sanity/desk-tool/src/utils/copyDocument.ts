import {isObject, omit} from 'lodash'

function copyReference(reference: Record<string, unknown>) {
  return Object.assign({}, reference)
}

function copyObject(
  object: Record<string, unknown>,
  options: {excludeReferences?: boolean; omit?: string[]} = {}
) {
  return Object.keys(object).reduce((copy, key) => {
    copy[key] = copyAny(object[key], options)
    return copy
  }, {})
}

function copyArray(array: unknown[], options: {excludeReferences?: boolean; omit?: string[]} = {}) {
  return array.map((item) => copyAny(item, options)).filter(Boolean)
}

const COPY_AS_IS = ['number', 'string', 'boolean']

function isReference(value) {
  return (
    value._type === 'reference' ||
    // should not happen as all references should have _type === 'reference'
    (!('_type' in value) && '_ref' in value)
  )
}

function copyAny(value: unknown, options: {excludeReferences?: boolean; omit?: string[]} = {}) {
  if (Array.isArray(value)) {
    return copyArray(value, options)
  }
  const type = typeof value
  if (COPY_AS_IS.includes(type)) {
    return value
  }
  if (!value) {
    return value
  }

  if (isObject(value)) {
    if (isReference(value)) {
      return options.excludeReferences ? undefined : copyReference(value as Record<string, unknown>)
    }

    return copyObject(value as Record<string, unknown>, options)
  }
  return value
}

export default function copyDocument(doc, options: {omit?: string[]} = {}) {
  const omitProps = ['_id'].concat(options.omit || [])
  return copyAny(omit(doc, omitProps), options)
}
