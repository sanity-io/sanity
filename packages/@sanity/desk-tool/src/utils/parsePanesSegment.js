/* eslint-disable import/prefer-default-export */
const panePattern = /^([a-z0-9_-]+),?({.*?})?(?:(;|$))/i

export function parsePanesSegment(str) {
  const chunks = []

  let buffer = str
  while (buffer.length) {
    const [match, id, paramsChunk] = buffer.match(panePattern) || []
    if (!match) {
      buffer = buffer.slice(1)
      continue
    }

    const params = paramsChunk && tryParseParams(paramsChunk)
    chunks.push({id, params})

    buffer = buffer.slice(match.length)
  }

  return chunks
}

function tryParseParams(json) {
  try {
    return JSON.parse(json)
  } catch (err) {
    // eslint-disable-next-line no-console
    console.warn(`Failed to parse parameters: ${err.message}`)
    return undefined
  }
}
