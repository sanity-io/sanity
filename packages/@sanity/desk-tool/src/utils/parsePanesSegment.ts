import {EMPTY_PARAMS} from '../constants'
import {exclusiveParams} from '../contexts/paneRouter'
import {RouterPaneGroup, RouterSplitPane} from '../types'

// old: authors;knut,{"template":"diaryEntry"}
// new: authors;knut,view=diff,eyJyZXYxIjoiYWJjMTIzIiwicmV2MiI6ImRlZjQ1NiJ9|latest-posts

const panePattern = /^([.a-z0-9_-]+),?({.*?})?(?:(;|$))/i
const isParam = (str: string) => /^[a-z0-9]+=[^=]+/i.test(str)
const isPayload = (str: string) =>
  /^(?:[A-Za-z0-9+/]{4})*(?:[A-Za-z0-9+/]{2}==|[A-Za-z0-9+/]{3}=)?$/.test(str)

type Truthy<T> = T extends false
  ? never
  : T extends ''
  ? never
  : T extends 0
  ? never
  : T extends 0n
  ? never
  : T extends null | undefined
  ? NonNullable<T>
  : T
const isTruthy = (Boolean as (t: unknown) => boolean) as <T>(t: T) => t is Truthy<T>

function parseChunks(chunks: string[], initial: RouterSplitPane): RouterSplitPane {
  return chunks.reduce(
    (pane, chunk) => {
      if (isParam(chunk)) {
        const key = chunk.slice(0, chunk.indexOf('='))
        const value = chunk.slice(key.length + 1)
        pane.params = {...pane.params, [key]: value}
      } else if (isPayload(chunk)) {
        pane.payload = tryParseBase64Payload(chunk)
      } else {
        // eslint-disable-next-line no-console
        console.warn('Unknown pane segment: %s - skipping', chunk)
      }

      return pane
    },
    {...initial, params: EMPTY_PARAMS, payload: undefined}
  )
}

function encodeChunks(pane: RouterSplitPane, index: number, group: RouterPaneGroup): string {
  const {payload, params = {}, id} = pane
  const sameAsFirst = index !== 0 && id === group[0].id
  const encodedPayload = typeof payload === 'undefined' ? undefined : btoa(JSON.stringify(payload))

  const encodedParams = Object.keys(params).reduce<string[]>((pairs, key) => {
    if (
      sameAsFirst &&
      index !== 0 &&
      !exclusiveParams.includes(key) &&
      group[0].params?.[key] === params[key]
    ) {
      return pairs
    }

    return params[key] ? [...pairs, `${key}=${params[key]}`] : pairs
  }, [])

  return (
    [sameAsFirst ? '' : id]
      .concat([encodedParams.length > 0 && encodedParams, encodedPayload].filter(isTruthy).flat())
      .join(',') || ','
  )
}

export function parsePanesSegment(str: string): RouterPaneGroup[] {
  if (str.indexOf(',{') !== -1) {
    return parseOldPanesSegment(str)
  }

  return str
    .split(';')
    .map((group) =>
      group
        .split('|')
        .map((segment) => {
          const [id, ...chunks] = segment.split(',')
          return parseChunks(chunks, {id})
        })
        .map((pane, i, siblings) => (pane.id ? pane : {...pane, id: siblings[0].id}))
    )
    .filter((group) => group.length > 0)
}

export function encodePanesSegment(panes: RouterPaneGroup[]): string {
  return (panes || [])
    .map((group) => group.map(encodeChunks).join('|'))
    .map(encodeURIComponent)
    .join(';')
}

export function parseOldPanesSegment(str: string): RouterPaneGroup[] {
  const chunks: RouterPaneGroup = []

  let buffer = str
  while (buffer.length) {
    const [match, id, payloadChunk] = buffer.match(panePattern) || []
    if (!match) {
      buffer = buffer.slice(1)
      continue
    }

    const payload = payloadChunk && tryParsePayload(payloadChunk)
    chunks.push({id, payload})

    buffer = buffer.slice(match.length)
  }

  return [chunks]
}

function tryParsePayload(json) {
  try {
    return JSON.parse(json)
  } catch (err) {
    // eslint-disable-next-line no-console
    console.warn(`Failed to parse parameters: ${err.message}`)
    return undefined
  }
}

function tryParseBase64Payload(data) {
  return data ? tryParsePayload(atob(data)) : undefined
}
