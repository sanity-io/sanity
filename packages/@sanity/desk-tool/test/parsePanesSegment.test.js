import {parsePanesSegment} from '../src/utils/parsePanesSegment'

test('parsePanesSegment()', () => {
  ;[
    'foo;bar,{"title":"foo;heisann"};baz;banana;hei',
    'foo;bar,{"nested":{"title":"foo;heisann","number":123}};baz;banana;hei',
    'fOo,{"invalid};',
    'fOo,{"invalid}',
    'foo,;bar',
    '  ',
    'foo;bar,{"inv};',
    'foo;bar,["foo"];baz'
  ].forEach(str => {
    expect(parsePanesSegment(str)).toMatchSnapshot(`Input: ${str}`)
  })
})
