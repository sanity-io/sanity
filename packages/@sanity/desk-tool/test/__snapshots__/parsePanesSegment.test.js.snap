// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parsePanesSegment(): Input:    1`] = `Array []`;

exports[`parsePanesSegment(): Input: fOo,{"invalid} 1`] = `
Array [
  Object {
    "id": "fOo",
    "params": undefined,
  },
]
`;

exports[`parsePanesSegment(): Input: fOo,{"invalid}; 1`] = `
Array [
  Object {
    "id": "fOo",
    "params": undefined,
  },
]
`;

exports[`parsePanesSegment(): Input: foo,;bar 1`] = `
Array [
  Object {
    "id": "foo",
    "params": undefined,
  },
  Object {
    "id": "bar",
    "params": undefined,
  },
]
`;

exports[`parsePanesSegment(): Input: foo;bar,["foo"];baz 1`] = `
Array [
  Object {
    "id": "foo",
    "params": undefined,
  },
  Object {
    "id": "baz",
    "params": undefined,
  },
]
`;

exports[`parsePanesSegment(): Input: foo;bar,{"inv}; 1`] = `
Array [
  Object {
    "id": "foo",
    "params": undefined,
  },
  Object {
    "id": "bar",
    "params": undefined,
  },
]
`;

exports[`parsePanesSegment(): Input: foo;bar,{"nested":{"title":"foo;heisann","number":123}};baz;banana;hei 1`] = `
Array [
  Object {
    "id": "foo",
    "params": undefined,
  },
  Object {
    "id": "bar",
    "params": Object {
      "nested": Object {
        "number": 123,
        "title": "foo;heisann",
      },
    },
  },
  Object {
    "id": "baz",
    "params": undefined,
  },
  Object {
    "id": "banana",
    "params": undefined,
  },
  Object {
    "id": "hei",
    "params": undefined,
  },
]
`;

exports[`parsePanesSegment(): Input: foo;bar,{"title":"foo;heisann"};baz;banana;hei 1`] = `
Array [
  Object {
    "id": "foo",
    "params": undefined,
  },
  Object {
    "id": "bar",
    "params": Object {
      "title": "foo;heisann",
    },
  },
  Object {
    "id": "baz",
    "params": undefined,
  },
  Object {
    "id": "banana",
    "params": undefined,
  },
  Object {
    "id": "hei",
    "params": undefined,
  },
]
`;
